internal/rpq/rpq.go

```diff
@@ -0,0 +1,141 @@
+// Package rpq implements a recurrent priority queue: a priority queue whose
+// elements can be returned multiple times, up to a specified quota,
+// prioritized according to how many times they have already been returned.
+package rpq
+
+import "container/heap"
+
+// A Queue is a recurrent priority queue of values of type V, optionally
+// indexed by keys of type K. Keys that are not the zero value of type K are
+// unique within a Queue.
+type Queue[K comparable, V any] struct {
+	pq    priorityQueue[K, V]
+	quota func() int
+}
+
+// An item is a key-value pair with an associated return count.
+type item[K comparable, V any] struct {
+	key   K
+	value V
+	count int
+}
+
+// New initializes a new Queue. Quota describes the minimum number of times an
+// item will be returned by Pop or PopN before it is removed from the Queue.
+func New[K comparable, V any](quota func() int) *Queue[K, V] {
+	return &Queue[K, V]{
+		pq:    makePriorityQueue[K, V](),
+		quota: quota,
+	}
+}
+
+// Push inserts a value into the Queue. If key is not the zero value of type K,
```

Comment:
Why should the zero value be special? Why not just always make it replace an existing value under key if there is one? (Also, it seems like that's actually what it does? So maybe this comment just needs to be updated.)

Reply:
This is a YAGNI holdover from an original concept keeping status messages and user messages in the same queue, which was problematic without a way to uniquely identify user messages to avoid delivering duplicates. But assigning unique IDs to user messages is a simple way to obviate that issue (as well as any utility of treating the zero value specially).